### ISSUE TITLE
Support `hints`

### ISSUE THREAD

**Issue by marandaneto** [getsentry/sentry-cocoa#2325]
### Description

This is needed for [https://github.com/getsentry/sentry-cocoa/issues/1460](<https://github.com/getsentry/sentry-cocoa/issues/1460>) and [https://github.com/getsentry/sentry-cocoa/issues/2650](<https://github.com/getsentry/sentry-cocoa/issues/2650>)
Also for [https://develop.sentry.dev/sdk/features/#http-client-errors](<https://develop.sentry.dev/sdk/features/#http-client-errors>) -> [PR](<https://github.com/getsentry/sentry-cocoa/pull/2308>)
`Hints` are part of the [Unified API](<https://develop.sentry.dev/sdk/unified-api/#hints>).
Hints are generally a `Map<String, Any>`, on Java leverage the data bag with a [typed class with helper methods](<https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Hint.java>).

**Comment by philipphofmann** [comment]
If we add an overload for `beforeSend` with a `hint` we can do this without a breaking change.

**Comment by philipphofmann** [comment]
On Java, we have [typed hints](https://github.com/getsentry/sentry-java/tree/beba01476ab71745b0cae821bed705d582374161/sentry/src/main/java/io/sentry/hints). I think that would be better than having a `Map<String, Any>`.

**Comment by marandaneto** [comment]
> On Java, we have [typed hints](https://github.com/getsentry/sentry-java/tree/beba01476ab71745b0cae821bed705d582374161/sentry/src/main/java/io/sentry/hints). I think that would be better than having a `Map<String, Any>`.

Yes, it used to be a `Map`, now it is a typed Hint, which is better.

**Comment by philipphofmann** [comment]
Hints must allow capturing stacktraces only for one event if capture stacktraces is disabled in the options, as requested in https://github.com/getsentry/sentry-cocoa/issues/3386.

**Comment by marcelofabri** [comment]
Thanks, @philipphofmann! Can you expand a little bit more on how much of a priority this is from Sentry's point of view and what should we expect in terms of timeline? 🙏

**Comment by kahest** [comment]
@marcelofabri We bumped the priority because we have more data points that `hints` would be a useful addition for some use cases. We don't plan to tackle this within the next few weeks at the moment so don't have a specific timeline, but we'd want to land this within the next months.

**Comment by sebastienboisvert** [comment]
Any updates on this? The (in)ability to add attachments to crash reports may be a deal-breaker for us.

**Comment by philipphofmann** [comment]
@sebastienboisvert, no sadly no updates yet. What is your use case for adding attachments to crash reports? Maybe there is already a working solution covering your use case.

**Comment by sebastienboisvert** [comment]
@philipphofmann as mentioned in some of the linked issues, we'd like to include text logs as an attachment as part of the crash report.

As an alternative I'm looking to see if there's a way to upload an attachment separately through the REST API (still reading through the docs).

**Comment by philipphofmann** [comment]
@sebastienboisvert, if your text logs aren't huge, you could consider adding breadcrumbs in before the SDK sends the crash event.

**Comment by oettam** [comment]
@philipphofmann in my case some files (binary format) imported in my iOS app could cause a crash. The ability to have such files attached to a crash report would massively help.

**Comment by bruno-garcia** [comment]
This blocks the ability to drop screenshots on BeforeSend as we do in other SDKs, including Unreal:
* https://github.com/getsentry/sentry-unreal/pull/849

**Comment by philprime** [comment]
## Research summary

> [!NOTE]
> This summary has been written by an agent on behalf of me for future reference

We reviewed the current SDK specification, the historical Unified API wording, the Java implementation, and the current Cocoa capture pipeline. This comment records the findings and a proposed project boundary. It is research, not yet an approved implementation design.

### Specification

The [historical Unified API hints page](https://develop.sentry.dev/sdk/unified-api/#hints) has been reorganized into the SDK foundations documentation. The core definition remains intentionally broad:

- A hint is transient, SDK-specific metadata about the origin of an event or breadcrumb.
- Capture and breadcrumb APIs may accept a hint.
- The same hint should be available to event processors and relevant callbacks such as `beforeSend` and `beforeBreadcrumb`.
- Hints are not part of the serialized event protocol.
- HTTP client integrations should provide original request and response objects as hints so users can filter or enrich events with context unavailable in the serialized event.

Relevant current specifications:

- [Hub and Scope concepts](https://develop.sentry.dev/sdk/foundations/state-management/hub-and-scope/)
- [Client hooks](https://develop.sentry.dev/sdk/foundations/client/hooks/)
- [HTTP client integrations](https://develop.sentry.dev/sdk/foundations/client/integrations/http-client/)

### Java reference implementation

Java uses a public final `io.sentry.Hint` data-bag class rather than an arbitrary object or subclass hierarchy:

- Arbitrary keyed values through `set`, `get`, `getAs`, and `remove`
- Typed retrieval with `getAs(key, Class<T>)`
- Per-event attachment operations
- Dedicated fields for screenshots, view hierarchies, thread dumps, tombstones, and replay recordings
- Convenience factories such as `Hint.withAttachment(...)`

Reference: [sentry-java `Hint.java`](https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Hint.java)

Java normalizes a missing hint to an empty `Hint` and passes the same mutable instance through the event pipeline:

1. Add scope attachments to the hint.
2. Run scoped event processors.
3. Run global event processors.
4. Run `beforeSend`.
5. Read final attachments from the hint while building the envelope.
6. Clear arbitrary source-object references before asynchronous transport work.

This makes the following possible without mutating a scope:

```java
options.setBeforeSend((event, hint) -> {
  hint.addAttachment(attachment);
  return event;
});
```

Java also places framework source objects in keyed hint entries, including HTTP requests and responses, logging records, Android activities/views, and navigation objects.

Java has a separate internal convention under the reserved `sentry:typeCheckHint` key. The stored object can implement internal interfaces such as `Cached`, `Retryable`, `SubmissionResult`, `Flushable`, and `AbnormalExit`. This lets the SDK carry transport and crash-processing state without adding those mechanics to the public `Hint` API. Cocoa should not copy this internal design where it already has equivalent mechanisms.

One important documentation discrepancy is that generic docs describe hints containing `event_id`, `originalException`, and `syntheticException`. Java does not automatically place the captured throwable or event ID in the public hint data bag. The throwable remains available on the event through `event.getThrowable()`, and the event ID remains on the event. This should inform the Cocoa API design rather than being copied blindly.

### Current Cocoa gaps

Cocoa currently has no hint type or propagation path.

The relevant callbacks accept only one argument:

```swift
beforeSend: (Event) -> Event?
beforeBreadcrumb: (Breadcrumb) -> Breadcrumb?
```

Capture calls flow through `SentrySDK`, `SentrySDKInternal`, `SentryHub`, and `SentryClient`, with matching Swift, Objective-C, `SentryObjC`, and `SentryObjCCompat` surfaces.

Attachments are currently assembled after event preparation and `beforeSend`. Scope attachments are collected and automatic attachment processors add screenshots or view hierarchies only after `beforeSend` has run. Consequently, `beforeSend` cannot inspect, add, replace, or remove the final per-event attachment set.

Cocoa also currently invokes `beforeSend` before its global event processors. Passing the same hint through processors and callbacks will require an explicit ordering decision in the design.

### Breaking-change analysis

Changing a callback from one parameter to two is a source-breaking public API change:

```swift
(Event) -> Event?
```

becomes:

```swift
(Event, Hint) -> Event?
```

The equivalent Objective-C block typedef change must also be treated as an ABI compatibility concern. Existing one-argument blocks must never be invoked as two-argument blocks.

The proposed rollout is:

- v9 default keeps the existing one-argument callback API.
- v9 can opt into the hint-aware API through the SPM package trait/compiler flag used for v10 API adoption.
- v10 exposes hint-aware callback signatures by default.

The design must keep Swift, Objective-C headers, `SentryObjC`, `SentryObjCCompat`, API snapshots, and tests aligned under both build modes.

Not every `beforeX` callback necessarily needs a hint. Initial scope should cover callbacks with a specification requirement or concrete use case, especially `beforeSend`, `beforeBreadcrumb`, event processors, and transaction handling where applicable.

### Proposed delivery scope

This is larger than the existing M estimate when treated as a complete Unified API feature. A reviewable delivery would likely use three Cocoa pull requests plus a separate sentry-docs pull request:

1. **Hint model and event pipeline**
   - Public, thread-safe hint data bag
   - Canonical hint-aware capture APIs
   - Propagation through SDK, hub, client, processors, and `beforeSend`
   - Original `NSError` and `NSException` access
   - v9 opt-in and v10-default callback behavior
   - Swift, Objective-C, and wrapper API coverage

2. **Per-event attachments**
   - Add, replace, and remove attachments through hints
   - Define scope-attachment merging
   - Resolve screenshot and view-hierarchy ordering
   - Cover fatal and stored-crash limitations
   - Build envelopes from the final post-callback attachment set

3. **Breadcrumb hints and integrations**
   - Hint-aware `addBreadcrumb` and `beforeBreadcrumb`
   - HTTP request and response source objects
   - Additional high-value integration source objects where justified

4. **Documentation**
   - Apple filtering and hints docs
   - Callback migration guidance
   - Per-event attachment examples
   - Supported source objects and stored-crash limitations

### Key risks to resolve in the design

- Source and binary compatibility of callback blocks
- Build-mode drift between v9 default, v9 opt-in, and v10
- Expensive screenshot/view-hierarchy work for events later dropped by `beforeSend`
- Retaining request, view, error, or exception objects longer than event preparation
- Stored crashes cannot preserve arbitrary in-memory hint objects across launches
- Avoiding an excessive overload matrix involving hint, scope, scope block, and thread options
- Compatibility-sensitive event processor APIs used by hybrid SDKs

### Tracking

A Linear project now tracks this effort and its known use cases:

- [Hints API [Apple]](https://linear.app/getsentry/project/hints-api-apple-b9b0ca2573a5)
- [getsentry/sentry-cocoa#2325: Support `hints`](https://github.com/getsentry/sentry-cocoa/issues/2325)
- [getsentry/sentry-cocoa#1460: Add attachments in `beforeSend`](https://github.com/getsentry/sentry-cocoa/issues/1460)
- [getsentry/sentry-cocoa#4180: `SentryEvent` does not return the original exception](https://github.com/getsentry/sentry-cocoa/issues/4180)
- [getsentry/sentry-cocoa#6540: Customize HTTP breadcrumbs with request information](https://github.com/getsentry/sentry-cocoa/issues/6540)

Per the SDK project-management playbook, this project requires a completed design document with risks and compatibility decisions before the first implementation work is reviewed.
